README.md
# [Master Typing](https://github.com/PythonDecorator)

<br />

![version](https://img.shields.io/badge/version-1.0.0-blue.svg)

--- 
![Thumnail.png](demo%2FThumnail.png)
## Table of Contents

* [Overview](#overview)
* [Demo](#demo)
* [Features](#features)
* [Documentation](#documentation)
* [Flow Chart](#flow-chart)
* [Converting to Executable](#converting-to-executable)
* [Controls](#Controls)
* [Licensing](#license)
* [Reporting Issues](#reporting-issues)
* [Technical Support or Questions](#technical-support-or-questions)
* [For Open Source](#For-open-source)
* [Social Media](#Social-media)

<br />

## Overview

Master Typing is a dynamic and user-friendly typing practice application developed in Python.

The purpose of this project is to help users improve their typing speed and accuracy through engaging and customizable
typing exercises. Whether you're a beginner looking to enhance your typing skills or a professional aiming to type
faster, Master Typing provides a platform for focused practice and skill refinement.

<br />

## Demo

![Master Typing Demo00.gif](demo%2FMaster%20Typing%20Demo00.gif)

- **Download the One file .exe file from the dist folder**
- **You don't need to install anything, just download and start typing**

<br />

## Features

> Some Main Features

1. ✅ `User-Friendly Interface`: Appealing interface that allows users to select duration, view their progress, and
   access various settings.

2. ✅ `Customizable Exercises`: Implemented a variety of typing exercises, including random sentences, quotes, and custom
   paragraphs, to cater to different practice preferences.

3. ✅ `Real-time Feedback`: Provide instant feedback to users as they type, marking correctly typed sentence.

4. ✅ `Timed Practice`: Incorporated timed typing sessions where users can challenge themselves to type as many words or
   characters as possible within a specified time frame.

5. ✅ `Word Count`: Keep track of users' correctly typed words, allowing them to monitor their progress and identify
   areas for improvement.

6. ✅ `Reset`: Provided options reset or restart, giving users control over their practice experience.

<br />

## Flow Chart
![Master Typing App FlowChart.png](files%2Fflow%20chart%2FMaster%20Typing%20App%20FlowChart.png)

<br />


## Documentation

This game was built based on the customtkinter and pygame documentation

<br />

## Converting to Executable

PyInstaller is a popular tool that allows you to convert Python scripts into standalone executable files for various
platforms, effectively creating desktop applications.

- You can use PyInstaller options to customize the behavior and appearance of the generated executable. Refer to the
  PyInstaller documentation for more information on available options.
- Keep in mind that PyInstaller generates a self-contained executable, but the size of the executable might be larger
  due
  to the inclusion of the Python interpreter and any dependencies, it's best to use venv to make sure only packages used
  in the
  project are included.
- Be sure to test the generated executable on the
  target platform to ensure everything works as expected.

<br />

> Install modules via `VENV`

```bash
$ virtualenv env
$ source env/bin/activate
$ pip3 install -r requirements.txt
```

<br />

> Create the .exe file

```bash
$ pyinstaller main.spec 
```

<br />


## License

This project is licensed under the MIT license. See also the attached LICENSE file.

<br />

## Reporting Issues

GitHub Issues is the official bug tracker for the Master Typing.

<br />

## Technical Support or Questions

If you have questions contact me `okpeamos.ao@gmail.com`

Make sure that you are using the latest version of the Master Typing. Check the CHANGELOG
Provide reproducible steps for the issue will shorten the time it takes for it to be fixed.

<br />

## For Open Source

The "Master Typing" project offers an excellent opportunity to create a useful and interactive typing practice
application using the CustomTkinter library in Python. By building this app, you'll enhance your skills in GUI
development, event handling, and user interface design. Whether you're looking to create a tool for personal improvement
or you want to contribute to the typing practice community, Master Typing allows you to combine technical proficiency
with a focus on helping users develop a valuable skill.

<br />

## Social Media

- Twitter: <https://twitter.com/AmosBrymo67154>
- Instagram: <https://www.instagram.com/pythondecorator>

<br />

---



CHANGELOG.md
# Change Log

## [1.0.0] 2023-04-07
### Initial Release
- Tic Tac Toe v1.0.0


data.py
import random
from wonderwords import RandomSentence


def get_sentence_to_type() -> tuple:
    # SENTENCES
    first = f"{s.sentence()[4:-1].capitalize()}, {s.bare_bone_sentence()[4:]}"
    second = random.choice(phrase_list)
    third = ", ".join(str(x) for x in adverbs_list_selected[:5])
    forth = f"{s.sentence()[4:-1].capitalize()}, {s.bare_bone_sentence()[4:]}"
    fifth = f"{s.sentence()[4:-1].capitalize()}, {s.bare_bone_sentence()[4:]}"
    sixth = random.choice(phrase_list)
    seventh = ", ".join(str(x) for x in adverbs_list_selected[5:])
    eight = f"{s.sentence()[4:-1].capitalize()}, {s.bare_bone_sentence()[4:]}"
    return first, second, third, forth, fifth, sixth, seventh, eight


adverbs = ['Now', 'Then', 'So', 'Very', 'Too', 'Never', 'Always', 'Ever', 'Seldom', 'Rarely', 'Gradually', 'Eventually',
           'Monthly', 'Weekly', 'Yearly', 'Annually', 'Quarterly', 'Much', 'Most', 'Quickly', 'Slowly', 'Incidentally',
           'Immediately', 'Simultaneously', 'Happily', 'Sadly', 'Frequently', 'Commonly', 'Sincerely', 'Faithfully',
           'Sweetly', 'Badly', 'Dearly', 'Patiently', 'Mostly', 'Silently', 'Willingly', 'Hardly', 'Often', 'Daily',
           'Occasionally', 'Regularly', 'Normally', 'Actually', 'Basically', 'Here', 'There', 'Yesterday', 'Today',
           'Tomorrow', 'Day after tomorrow', 'Day before yesterday', 'Tonight', 'Extremely', 'Exceedingly', 'Arguably',
           'Comparatively', 'Consecutively', 'Honestly', 'Truthfully', 'Lovingly', 'Perfectly', 'Highly', 'Likely',
           'Nearly', 'Barely', 'Far', 'Least', 'Deeply', 'Fully', 'Completely', 'Casually', 'Tastefully', 'Madly',
           'Purely', 'Privately', 'Publicly', 'Towards', 'Inside', 'Outside', 'Upstairs', 'Downstairs', 'Uphill',
           'Downhill', 'Everywhere', 'Somewhere', 'Nowhere', 'Anywhere', 'Eagerly', 'Beautifully', 'Proudly',
           'Elegantly', 'Confidently', 'Incessantly', 'Boldly', 'Carefully', 'Cautiously', 'Carelessly', 'Easily',
           'Awkwardly', 'Nearby', 'Cheerfully', 'Abruptly', 'Late', 'Everyday', 'Soon', 'Coldly', 'Angrily',
           'Curiously', 'Noisily', 'Loudly', 'Earnestly', 'Interestingly', 'Readily', 'Vaguely', 'Unwillingly',
           'Obediently', 'Rapidly', 'Continuously', 'Consciously', 'Instinctively', 'Boldly', 'Brightly', 'Cunningly',
           'Suitably', 'Appropriately', 'Currently', 'Doubtfully', 'Ambiguously', 'Momentarily', 'Early', 'Last week',
           'Last month', 'Last year', 'Later', 'Northwards', 'Southwards', 'Eastwards', 'Westwards', 'Forward',
           'Backwards', 'Away', 'In', 'Out', 'Under', 'Below', 'Above', 'Abroad', 'North', 'South', 'West', 'East',
           'Southeast', 'Southwest', 'Northeast', 'Northwest', 'Up', 'Down', 'Underneath', 'Gently', 'Superficially',
           'Supremely', 'Adequately', 'Comfortably', 'Conveniently', 'Generously', 'Briefly', 'Accidentally', 'Tightly',
           'Fearfully', 'Gracefully', 'Graciously', 'Busily', 'Randomly', 'Joyously', 'Mysteriously', 'Joyfully',
           'Poorly', 'Repeatedly', 'Seriously', 'Smoothly', 'Promptly', 'Roughly', 'Successfully', 'Sufficiently',
           'Skillfully', 'Sceptically', 'Differently', 'Physically', 'Psychologically', 'Logically', 'Analytically',
           'Graphically', 'Loosely', 'Fiercely', 'Unexpectedly', 'Tactfully', 'Lazily', 'Tremendously', 'Vicariously',
           'Vividly', 'Cleverly', 'Victoriously', 'Widely', 'Well', 'Purposefully', 'Wisely', 'Properly', 'Sickly',
           'Legally', 'Nicely', 'Legibly', 'Thoroughly', 'Shortly', 'Simply', 'Tidily', 'Necessarily', 'Tenaciously',
           'Strongly', 'Humbly', 'Consequently', 'Similarly', 'Unlikely', 'Possibly', 'Probably']

phrase = """Yeah, right. You're kidding. You're pulling my leg.
That's a bit of an exaggeration. He's stretching the truth.
He's not telling the whole truth. Being economical with the truth.
His story is fishy. That's an outright lie. That's a pack of lies.
They got off on the wrong foot. He got on the teacher's bad side.
She took offense at his comment. He has a chip on his shoulder.
He has a short fuse. he dissed my mother. Got his nose out of joint.
It's as light as a feather. As dry as a bone, as @strong as an ox.
It's as flat as a pancake. He's as mad as a hornet.
It's as old as the hills. It's as quick as lightning.
She's as sick as a dog. They're as different as night and day.
She's as stubborn as a mule. He's as proud as a peacock.
She's as white as a sheet. It's as solid as a rock.
It's as clear as mud, "I couldn't wait to get back home."
What's the matter?. What's wrong?. Are you all right?
You look a bit down. Is there anything I can do to help?
Cheer up! Chin up! It's not so bad. Everything will be OK.
Look on the bright side. It's not the end of the world.
It's fascinating. I couldn't tear myself away.
I couldn't put it down. I was so into it, I lost track of time.
It does nothing for me. I was bored to tears. It's intriguing.
I was bored to death. I was dying of boredom.
It's about as exciting as watching paint dry. I'm absolutely sure.
I'm positive that it will work. I'm a hundred percent certain.
Chances are that (this will probably happen).
I seriously doubt it. I don't think so. Probably not.
It's not very likely. There's not much chance of that.
I'd be very surprised if that happened. I wouldn't bet on it
You look nice / You look amazing! What a beautiful [@necklace]!
I like your [shirt/shoes/haircut/etc.]. The color is perfect!
You're a fantastic cook. My compliments to the chef!
3. What a nice apartment! 4. You have a beautiful home.
He's/She's so cute! Your kids are a lot of fun.
Lucky you! That was a stroke of luck "sudden event of good luck".
Some people have all the luck. As luck would have it (by chance)
He's down on his luck. "having a long period of difficulty"
I'm debating between [option A & B], I'm yet to make up my mind.
I'm on the fence, "I'm in the middle, I don't know what to decide)."
I'll take that into consideration. On the other hand…
I'm having second thoughts "I'm reconsidering my decision"
I changed my mind. He convinced/persuaded me to…
Looking back, I know it was the right decision. It's up to you
Could you give me a minute? (informal) Hang on a sec / Just a sec.
Hold on. Let me see/think. I'll be right with you. Bear with me.
That'll have to wait. Be patient. Not so fast! Hold your horses!
She was born to dance. He's a natural. He could do it in his sleep.
15. He knows it inside out. Call +1 903 395 3804
She knows [New York] like the back of her hand.
She's a walking encyclopedia of [philosophy]. In a class of his own.
He's the best in the business. She's very gifted.
He's a [chemistry] whiz. No comment. Check pink1398@gmail.com
I'm not at liberty to say to give out such information.
Wait and see, "You will discover the answer later"
Let me get back to you. (I will give you the answer later)
I'm sorry, that's confidential. I'm sorry, that's personal.
I'd rather not talk about it. It's none of your business.
1. Mind your own business. 2. Why do you want to know?
""".split("\n ")

s = RandomSentence()
adverbs_list_selected = [random.choice(adverbs) for x in range(10)]
phrase_list = phrase[0].split("\n")

tip_1 = "Use the correct starting position.\n\nWhen practicing your typing skills, " \
        "it's important to use proper hand placement. To start, keep your fingers " \
        "positioned over the home row keys (left hand over the A, S, D, and F keys, " \
        "and the right hand over the J, K, L, and ; keys).\n\nWith your thumbs hovering over the space bar. " \
        "From here, you can move your fingers slightly to reach neighboring keys."

tip_2 = "Don't look down your hands.\n\nInstead of looking down at your hands, focus on your screen. " \
        "This can be difficult at first, especially if you have not yet mastered the exact placement of the keys." \
        "\n\nHowever, looking at the screen will help improve your accuracy because you will be able to catch your " \
        "typos as they occur. You'll also begin to memorize the placement of the keys, so you'll be able to " \
        "type more quickly as you practice."


styling.py
# FONTS
ENTRY_FONT = ("Times New Roman", 19, "normal")
TIMER_FONT = ("Lucida Sans Typewriter", 160, "bold")
TIMER_LABEL_FONT = ("Roboto Slab", 25, "normal")
WORD_COUNT_FONT = ("Lucida Sans Typewriter", 100, "bold")
WORD_COUNT_LABEL_FONT = TIMER_LABEL_FONT
TIPS_FONT = ("Times New Roman", 15, "normal")
CANVAS_TIP_FONT = ("Times New Roman", 30, "normal")
BOTTOM_FONT = TIPS_FONT
LOGO_FONT = ("Roboto Slab", 40, "bold")
START_FONT = TIMER_LABEL_FONT

# COLORS
BG_COLOR = "#0A3D51"
FRAME_COLOR = "#259DD8"
TYPE_TEXT_COLOR = "#FFFFFF"
ENTRY_TEXT_COLOR = TYPE_TEXT_COLOR
SCROLL_COLOR = "#000000"
EDIT_COLOR = BG_COLOR
TIMER_COLOR = FRAME_COLOR
WORD_COUNT_COLOR = TIMER_COLOR
BOTTOM_TEXT_COLOR = TIMER_COLOR
START_BTN_COLOR = SCROLL_COLOR
MENU_BTN_COLOR = "#0980AF"
MENU_HOVER_COLOR = "#09698E"
STARTED_BG_COLOR = "#890906"
TIME_COUNT_COLOR = "#06602C"
RESTART_HOVER_COLOR = "#02913F"
TIPS_COLOR = SCROLL_COLOR


main.py
import customtkinter as ctk
from styling import *
from PIL import Image
from data import tip_1, tip_2, get_sentence_to_type
import pygame
import sys
import os


class App(ctk.CTk):
    def __init__(self):
        super().__init__(fg_color=BG_COLOR)
        self.title(" Master Typing")
        self.resizable(False, False)
        self.iconbitmap(self.resource_path("files/images/logo.ico"))

        app_width = 1000
        app_height = 700
        screen_width = self.winfo_screenwidth()
        screen_height = self.winfo_screenheight()
        x_cor = int((screen_width / 2) - (app_width / 2))
        y_cor = int((screen_height / 2) - (app_height / 2))
        self.geometry("{}x{}+{}+{}".format(app_width, app_height, x_cor, y_cor))

        # MUSIC
        pygame.init()
        self.bg_music = pygame.mixer.Sound(self.resource_path("files/sound/not_started.mp3"))
        self.bg_music.set_volume(0.5)
        self.bg_music.play(loops=-1)
        self.typing_music = pygame.mixer.Sound(self.resource_path("files/sound/typing.mp3"))
        self.logo_effect_sound = pygame.mixer.Sound(self.resource_path("files/sound/sound effect.mp3"))
        self.logo_effect_sound.set_volume(0.1)

        # VARIABLES
        self.start_btn_var = ctk.StringVar(value="Start")
        self.timer_int_var = ctk.IntVar(value=60)
        self.timer_menu_var = ctk.StringVar(value="60 sec")

        # LAYOUT DESIGN
        # LOGO FRAME
        LogoFrame(self, self.resource_path)

        #  WORD COUNT FRAME
        self.word_count = WordCountFrame(self)

        # TIMER FRAME
        self.timer = TimerFrame(self)

        # MENU AND START BUTTON FRAME
        self.start_btn = MenuAndStartButtonFrame(self)

        #  SCROLLABLE FRAME
        self.scrollable_frame = ScrollableFrame(self)

        # # LEFT CANVAS FRAME
        LeftTipFrame(self)

        # RIGHT CANVAS FRAME
        RightTipFrame(self)

        # BOTTOM TEXT LABEL
        bottom_text = ctk.CTkLabel(self, text="Designed by Amos @PythonDecorator",
                                   text_color=BOTTOM_TEXT_COLOR, font=BOTTOM_FONT)
        bottom_text.place(relx=0.5, rely=0.976, anchor="center")

        self.start_btn_var.trace_add("write", self.update_music)
        self.start_btn_var.trace_add("write", self.timer.update_timer)
        self.timer_int_var.trace_add("write", self.detect_time_up)

    def resource_path(self, relative_path):
        """ Get absolute path to resource, works for dev and for PyInstaller """
        base_path = getattr(sys, '_MEIPASS', os.path.dirname(os.path.abspath(__file__)))
        return os.path.join(base_path, relative_path)

    def detect_time_up(self, *args):
        time_left = self.timer_int_var.get()

        if self.start_btn_var.get() == "Started":
            if time_left > 30:
                self.timer.canvas.itemconfigure(self.timer.timer, fill=TIME_COUNT_COLOR)
            else:
                self.timer.canvas.itemconfigure(self.timer.timer, fill=STARTED_BG_COLOR)

        if not self.timer_int_var.get():
            for label, entry in self.scrollable_frame.label_and_entry_list:
                entry.configure(state="disabled")

            self.start_btn_var.set(value="Restart")
            self.timer_int_var.set(value=int(self.timer_menu_var.get().split(" ")[0]))

            self.timer.canvas.itemconfigure(self.timer.timer, fill=STARTED_BG_COLOR)
            self.start_btn.start_btn.configure(text=self.start_btn_var.get(), fg_color=RESTART_HOVER_COLOR,
                                               state="normal", hover_color=TIME_COUNT_COLOR)

    def update_music(self, *args):
        if self.start_btn_var.get() == "Started":
            self.start_btn_var.set(value="Started")
            self.bg_music.stop()
            self.typing_music.play(loops=-1)
        else:
            self.bg_music.play(loops=-1)
            self.typing_music.stop()


class LogoFrame(ctk.CTkFrame):
    def __init__(self, parent: App, get_image_func):
        super().__init__(master=parent, fg_color=SCROLL_COLOR, corner_radius=0)
        self.place(relx=0.2, rely=0.024, relwidth=0.7, relheight=0.18)
        self.parent = parent
        self.path = get_image_func

        self.ctk_image_list = []
        self.logo_image_frames = self.get_images()
        self.current_frame = 0
        self.last_frame = len(self.logo_image_frames) - 1

        self.logo_label = ctk.CTkLabel(self, text="")
        self.logo_label.place(relx=0, rely=0.15, relwidth=0.87, relheight=0.98)

        self.parent.start_btn_var.trace_add("write", self.update_animation)
        self.parent.start_btn_var.trace_add("write", self.play_logo_sound)
        self.play_logo_sound()
        self.update_animation()

    def play_logo_sound(self, *args):
        if self.parent.start_btn_var.get() != "Started":
            self.parent.logo_effect_sound.play()
            self.after(2000, self.play_logo_sound)
        else:
            self.parent.logo_effect_sound.stop()

    def update_animation(self, *args):
        self.logo_label.configure(image=self.logo_image_frames[self.current_frame])

        if self.parent.start_btn_var.get() != "Started":
            if self.current_frame < self.last_frame:
                self.current_frame += 1
                self.after(40, self.update_animation)
            else:
                self.current_frame = 0
                self.after(40, self.update_animation)
        else:
            self.logo_label.configure(image=self.logo_image_frames[49])
            self.current_frame = 0

    def get_images(self):
        dark = [Image.open(self.path(f"files/logo/logo/{n}.png")) for n in range(1, 51)]
        light = [Image.open(self.path(f"files/logo/logo/{n}.png")) for n in range(1, 51)]
        self.ctk_image_list = [ctk.CTkImage(light_image=light[n], dark_image=dark[n], size=(550, 109))
                               for n in range(len(dark))]
        return self.ctk_image_list


class WordCountFrame(ctk.CTkFrame):
    def __init__(self, parent: App):
        super().__init__(master=parent, fg_color=FRAME_COLOR, corner_radius=0)
        self.place(relx=0.022, rely=0.024, relwidth=0.2, relheight=0.18)
        self.parent = parent

        self.canvas = ctk.CTkCanvas(self, bg=FRAME_COLOR, highlightthickness=0)
        self.canvas.create_line((90, 160, 400, 160), fill=SCROLL_COLOR, width=10)
        self.canvas.pack(expand=True, fill="both")

        self.typed_words = self.canvas.create_text((250, 90), text="0",
                                                   fill=SCROLL_COLOR, font=WORD_COUNT_FONT)
        self.total_words = self.canvas.create_text((250, 225), fill=SCROLL_COLOR, font=WORD_COUNT_FONT)


class TimerFrame(ctk.CTkFrame):
    def __init__(self, parent: App):
        super().__init__(master=parent, fg_color=FRAME_COLOR, corner_radius=0)
        self.place(relx=0.779, rely=0.024, relwidth=0.2, relheight=0.18)
        self.parent = parent

        self.canvas = ctk.CTkCanvas(self, bg=FRAME_COLOR, highlightthickness=0)
        self.canvas.pack(expand=True, fill="both")

        self.timer = self.canvas.create_text((250, 160), text=str(self.parent.timer_int_var.get()),
                                             fill=SCROLL_COLOR, font=TIMER_FONT)

    def update_timer(self, *args):
        time_left = self.parent.timer_int_var.get()
        time_left -= 1
        self.parent.timer_int_var.set(value=time_left)
        self.canvas.itemconfigure(self.timer, text=str(time_left))

        if time_left > 0:
            self.after(1000, self.update_timer)


class MenuAndStartButtonFrame(ctk.CTkFrame):
    def __init__(self, parent: App):
        super().__init__(master=parent, fg_color=BG_COLOR, corner_radius=0)
        self.place(relx=0.022, rely=0.2128, relwidth=0.958, relheight=0.04)
        self.parent = parent

        # TIMER LABEL
        self.timer_label = ctk.CTkLabel(self, text="Timer", fg_color=BG_COLOR, font=TIMER_LABEL_FONT,
                                        text_color=FRAME_COLOR)
        self.timer_label.place(relx=0.735, rely=0.01, relwidth=0.2, relheight=0.9)

        # WORD COUNT LABEL
        self.word_count_label = ctk.CTkLabel(self, text="Word Count", fg_color=BG_COLOR, font=TIMER_LABEL_FONT,
                                             text_color=FRAME_COLOR)
        self.word_count_label.pack(side="left", padx=25)

        # TIMER MENU
        self.timer_menu = ctk.CTkOptionMenu(self, fg_color=FRAME_COLOR, text_color=SCROLL_COLOR, font=TIPS_FONT,
                                            values=["60 sec", "90 sec", "120 sec", "180 sec", "250 sec", "300 sec",
                                                    "400 sec", "500 sec"], dropdown_fg_color=BG_COLOR,
                                            width=80, button_hover_color=MENU_HOVER_COLOR, button_color=MENU_BTN_COLOR,
                                            dropdown_text_color=FRAME_COLOR, dropdown_hover_color=MENU_HOVER_COLOR,
                                            command=self.update_timer_seconds, variable=self.parent.timer_menu_var)
        self.timer_menu.pack(side="right", padx=10)

        # START BUTTON
        self.start_btn = ctk.CTkButton(self, text=self.parent.start_btn_var.get(), font=TIMER_LABEL_FONT,
                                       fg_color=FRAME_COLOR, text_color=START_BTN_COLOR, hover_color=MENU_HOVER_COLOR,
                                       command=self.update_button)
        self.start_btn.place(relx=0.5, rely=0.5, relwidth=0.2, relheight=1, anchor="center")

    def update_timer_seconds(self, *args):
        selected_time = int(self.parent.timer_menu_var.get().split(" ")[0])
        if self.parent.start_btn_var.get() != "Started":
            self.parent.timer_int_var.set(value=selected_time)
            self.parent.timer.canvas.itemconfigure(self.parent.timer.timer, text=selected_time, fill=SCROLL_COLOR)

    def update_button(self, *args):
        if self.parent.start_btn_var.get() == "Start" or "Restart":
            self.parent.start_btn_var.set(value="Started")
            self.start_btn.configure(text=self.parent.start_btn_var.get(), fg_color=STARTED_BG_COLOR, state="disabled")

        if self.parent.start_btn_var.get() == "Started":
            self.parent.scrollable_frame.sentence_list = get_sentence_to_type()
            word_count_list = len([word for sentence in self.parent.scrollable_frame.sentence_list
                                   for word in sentence.split(" ")])

            for index, (label, entry) in enumerate(self.parent.scrollable_frame.label_and_entry_list):
                label.configure(text=f" {self.parent.scrollable_frame.sentence_list[index]}",
                                text_color=TYPE_TEXT_COLOR)
                entry.configure(state="normal")
                entry.delete(0, 'end')

            self.parent.word_count.canvas.itemconfigure(self.parent.word_count.total_words, text=f"{word_count_list}")


class ScrollableFrame(ctk.CTkScrollableFrame):
    def __init__(self, parent: App):
        super().__init__(master=parent, fg_color=SCROLL_COLOR, corner_radius=0,
                         scrollbar_button_hover_color=MENU_HOVER_COLOR, scrollbar_button_color=BG_COLOR)
        self.place(relx=0.218, rely=0.26, relwidth=0.56, relheight=0.696)
        self.parent = parent

        self.entry_var_list = [ctk.StringVar() for _ in range(8)]

        self.label_and_entry_list = []
        self.sentence_list = get_sentence_to_type()
        word_count_list = [word for sentence in self.sentence_list for word in sentence.split(" ")]
        word_count = len(word_count_list)
        self.parent.word_count.canvas.itemconfigure(self.parent.word_count.total_words, text=f"{word_count}")

        for n in range(8):
            label = ctk.CTkLabel(self, text=f" {self.sentence_list[n]}", fg_color=BG_COLOR, font=ENTRY_FONT, anchor="w")
            entry = ctk.CTkEntry(self, fg_color=BG_COLOR, font=ENTRY_FONT, border_color=MENU_HOVER_COLOR,
                                 textvariable=self.entry_var_list[n])
            self.label_and_entry_list.append((label, entry))

        for label, entry in self.label_and_entry_list:
            label.pack(expand=True, fill="both", pady=10, padx=10)
            entry.pack(expand=True, fill="both", pady=10, padx=10)

        for entry_var in self.entry_var_list:
            entry_var.trace_add("write", self.update_word_count)

    def update_word_count(self, *args):
        score = 0

        correct_sentence_list = [label.cget("text")[1:] for label, entry in self.label_and_entry_list]
        sentence_typed_list = [entry_var.get() for entry_var in self.entry_var_list]

        correct_words_list = [word.split(" ") for word in correct_sentence_list]
        words_typed_list = [entry_var.get().split(" ") for entry_var in self.entry_var_list]

        for n in range(len(correct_sentence_list)):
            if correct_sentence_list[n] == sentence_typed_list[n]:
                self.label_and_entry_list[n][0].configure(text=f" {correct_sentence_list[n]} ✅",
                                                          text_color=TIME_COUNT_COLOR)

            current_input_word_list = correct_words_list[n]
            current_input_typed = words_typed_list[n]
            for x in range(len(current_input_typed) - 1):
                if current_input_typed[x] == current_input_word_list[x]:
                    score += 1

        self.parent.word_count.canvas.itemconfigure(self.parent.word_count.typed_words,
                                                    text=score)


class LeftTipFrame(ctk.CTkFrame):
    def __init__(self, parent: App):
        super().__init__(master=parent, fg_color=FRAME_COLOR, corner_radius=0)
        self.place(relx=0.022, rely=0.26, relwidth=0.2, relheight=0.695)
        self.parent = parent

        self.canvas = ctk.CTkCanvas(self, bg=FRAME_COLOR, highlightthickness=0)
        self.canvas.pack(expand=True, fill="both")

        self.canvas.create_text((40, 40), text=tip_1, fill=TIPS_COLOR, font=CANVAS_TIP_FONT, width=430, anchor="nw")


class RightTipFrame(ctk.CTkFrame):
    def __init__(self, parent: App):
        super().__init__(master=parent, fg_color=FRAME_COLOR, corner_radius=0)
        self.place(relx=0.779, rely=0.26, relwidth=0.2, relheight=0.695)
        self.parent = parent

        self.canvas = ctk.CTkCanvas(self, bg=FRAME_COLOR, highlightthickness=0)
        self.canvas.pack(expand=True, fill="both")

        self.canvas.create_text((40, 40), text=tip_2, fill=TIPS_COLOR, font=CANVAS_TIP_FONT, width=430, anchor="nw")


if __name__ == "__main__":
    app = App()
    app.mainloop()
